core/src/main/java/com/linecorp/armeria/client/limit/DefaultConcurrencyLimit.java

```diff
@@ -0,0 +1,182 @@
+        this.predicate = predicate;
+        this.maxConcurrency = maxConcurrency;
+        this.maxPendingAcquisitions = maxPendingAcquisitions;
+        this.timeoutMillis = timeoutMillis;
+    }
+
+    @VisibleForTesting
+    int acquiredPermits() {
+        return acquiredPermits.get();
+    }
+
+    @VisibleForTesting
+    int availablePermits() {
+        return maxConcurrency - acquiredPermits.get();
+    }
+
+    @Override
+    public CompletableFuture<SafeCloseable> acquire(ClientRequestContext ctx) {
+        if (!predicate.test(ctx)) {
+            return noLimitFuture;
+        }
+
+        if (pendingAcquisitions.isEmpty()) {
+            if (acquiredPermits.incrementAndGet() <= maxConcurrency) {
+                return CompletableFuture.completedFuture(new Permit());
+            }
+            acquiredPermits.decrementAndGet();
+        }
+
+        if (numPendingAcquisitions.incrementAndGet() > maxPendingAcquisitions) {
+            numPendingAcquisitions.decrementAndGet();
+            return CompletableFutures.exceptionallyCompletedFuture(
+                    UnprocessedRequestException.of(TooManyPendingAcquisitionsException.get()));
+        }
+
+        final CompletableFuture<SafeCloseable> future = new CompletableFuture<>();
+        final PendingAcquisition pendingAcquisition = new PendingAcquisition(ctx, future);
+        pendingAcquisitions.add(pendingAcquisition);
+        // Call drain to check if there's available permits in the meantime.
+        drain();
+        return future;
+    }
+
+    void drain() {
+        while (!pendingAcquisitions.isEmpty()) {
+            final int currentAcquiredPermits = acquiredPermits.get();
+            if (currentAcquiredPermits >= maxConcurrency) {
+                break;
+            }
+
+            if (acquiredPermits.compareAndSet(currentAcquiredPermits, currentAcquiredPermits + 1)) {
+                final PendingAcquisition task = pendingAcquisitions.poll();
+                if (task == null) {
+                    acquiredPermits.decrementAndGet();
+                    if (!pendingAcquisitions.isEmpty()) {
+                        // Another task might have been added to the queue while acquiredPermits reached
+                        // at its limit.
+                        continue;
+                    } else {
+                        break;
+                    }
+                }
+
+                task.run();
+                numPendingAcquisitions.decrementAndGet();
+            }
+        }
+    }
+
+    private final class PendingAcquisition implements Runnable {
+
+        private final CompletableFuture<SafeCloseable> future;
+        @Nullable
+        private final ScheduledFuture<?> timeoutFuture;
+
+        PendingAcquisition(ClientRequestContext ctx, CompletableFuture<SafeCloseable> future) {
+            this.future = future;
+            if (timeoutMillis != 0) {
+                timeoutFuture = ctx.eventLoop().withoutContext().schedule(
```

Comment:
Looks like still pending count isn't updated until drain for timeouts. Is it ok? Aren't there cases it takes reasonable time until the next drain?